Blog post: Securely direct traffic from IBM Cloud Kubernetes Service Ingress to Istio Ingress Gateway

vadimeisenbergibm opened the pull request:
Please provide a description for what this PR is for.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[X ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

istio-testing commented:
@vadimeisenbergibm: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio.io | f44fecf624197348466f0c3b3bd2ca0984713b3e | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio.io/5098/lint_istio.io/481) | `/test lint_istio.io`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

rvennam commented on content/en/docs/examples/alb-nlb/index.md:
```diff
@@ -0,0 +1,363 @@
+
+In this example you will configure the IKS Ingress ALB to direct traffic to the services inside an Istio service mesh
+through the Istio ingress gateway, while using mutual TLS between ALB and the gateway.
+The traffic to the services without Istio sidecar can continue to flow as before, directly from ALB.
```
I think a diagram can help here. Maybe something like this?
<img width="591" alt="image" src="https://user-images.githubusercontent.com/5502967/66315731-5b180880-e8e4-11e9-886e-763356b616ab.png">

vadimeisenbergibm replied:
yep, let me create a diagram

linsun replied:
I think this diagram isn't entirely correct.  it would be 2 flows:
1. Client - ALB - istio ingress gateway - pod
2. Client - NLB - istio ingress gateway - pod

Agree?

rvennam replied:
Right, the diagram needs to be modified a bit to show both flows

vadimeisenbergibm replied:
right, I will change the diagram according to the example

linsun commented on content/en/docs/examples/alb-nlb/index.md:
```diff
@@ -0,0 +1,363 @@
+[Ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/) resource with
+[ALB-specific annotations](https://cloud.ibm.com/docs/containers?topic=containers-ingress_annotation).
+
+When you start using Istio in your IKS cluster, it is recommended that you use the
```
did we explain what IKS stands for earlier?

vadimeisenbergibm replied:
I will add a link to IKS.

linsun commented on content/en/docs/examples/alb-nlb/index.md:
```diff
@@ -0,0 +1,363 @@
+
+When you start using Istio in your IKS cluster, it is recommended that you use the
+[Istio Ingress Gateway](/docs/tasks/traffic-management/ingress/ingress-control/) instead of using Kubernetes Ingress
+resource. One of the main reasons to use Istio ingress gateway is the fact the ALB provided by IKS will not be able to
```
not be able to communicate *directly*.... as you can do it via istio ingress gateway or add an envoy sidecar to ALB.

vadimeisenbergibm replied:
right, I will add _directly_.

linsun commented on content/en/docs/examples/alb-nlb/index.md:
```diff
@@ -0,0 +1,363 @@
+    {{< text bash >}}
+    $ kubectl create -n istio-system secret tls istio-ingressgateway-certs --key nlb_certs/tls.key --cert trusted.crt
+    $ kubectl create -n istio-system secret generic istio-ingressgateway-ca-certs --from-file=trustid-x3-root.pem
+    $ kubectl create secret generic alb-certs -n istio-system --from-file=trusted.crt --from-file=alb_certs/client.crt --from-file=alb_certs/client.key
```
who is using alb-certs?

vadimeisenbergibm replied:
The ALB.

linsun replied:
u mean istio ingress gateway?

vadimeisenbergibm replied:
no, the ALB uses alb-certs to communicate with the Istio ingress gateway via mutual TLS

vadimeisenbergibm replied:
Yeah, the issue is complex. Maybe the summary will clarify it, added in https://github.com/istio/istio.io/pull/5098/commits/e6307814cc8f1a640f6ad8461994f9573c018f46

linsun commented on content/en/docs/examples/alb-nlb/index.md:
```diff
@@ -0,0 +1,363 @@
+---
+title: Securely direct traffic from IBM Cloud Kubernetes Service Ingress to Istio Ingress Gateway
```
This title is a bit too long, especially if you view it from preview, it is shown on the left panel.

I think it would be good to have 2 blogs.  I found it confusing that NLB and ALB and talked together while users configure NLB to istio ingress gateway and/and ALB to istio ingress gateway.  thoughts?

vadimeisenbergibm replied:
@rvennam ^ what do you think?

frankbu commented:
We need to decide where platform-specific examples belong. We could have sub folders in the examples section, e.g., `examples/ibmcloud/`, `examples/gke/`, and only have generic examples in the top level `examples/` folder. Or, we can decide that there are no examples that are for a single platform on istio.io, in which case we could make this a blog entry instead. As a general rule, we don't want want to clutter up istio.io with a lot of docs that only apply to one platform or another.

I'm adding this to the agenda for the next doc WG call, so we can decide on a policy.

istio-testing commented:
@vadimeisenbergibm: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio.io | bd83adadf1f34232bd4bcc482c70348b81cd1ad7 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio.io/5098/lint_istio.io/637) | `/test lint_istio.io`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@vadimeisenbergibm: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio.io | d6b979c92ddbdcc9364badf31ede4855fd04f816 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio.io/5098/lint_istio.io/646) | `/test lint_istio.io`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

vadimeisenbergibm commented:
@linsun @rvennam @frankbu I rewrote the example as a blog post and did the following changes - please read the introduction only, the rest is just technical details.

1. Removed mentioning NLB (the Istio Ingress Gateway is the NLB in this context).
1. Explained the involved DNS domain names and how they are related to certificates/keys
1. Added a diagram

Now the blog post should be more clear. Your comments are welcome.

istio-testing commented:
@vadimeisenbergibm: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio.io | f1b42633f16c558b76d8318b129f388ac17325c2 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio.io/5098/lint_istio.io/647) | `/test lint_istio.io`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

linsun commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,389 @@
+    deployment.apps/httpbin created
+    {{< /text >}}
+
+1.  Require mutual TLS for the traffic to the services in
```
Require -> Enforce

vadimeisenbergibm replied:
Done in https://github.com/istio/istio.io/pull/5098/commits/c0eeaba55db65f72e083eb49f7773f69c7a4f800.

linsun commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,389 @@
+    IKS 37 days before they expire.
+    You will have to recreate the secrets by rerunning the instructions of this section every time the secrets provided
+    by IKS are updated. You may want to use scripts or operators to automate this and keep the
+    secrets in sync.
```
Does SDS ingress gateway mode make this any  easier?

vadimeisenbergibm replied:
SDS ingress gateway will not help with copying since it can only access the secrets in the namespace of the gateway. In this case, the secrets are generated by IKS in the default namespace, while the gateway is in `istio-system`.

linsun replied:
Agreed, thanks for clarifying it.

linsun commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,389 @@
+    $ kubectl create secret generic alb-certs -n istio-system --from-file=trusted.crt --from-file=alb_certs/client.crt --from-file=alb_certs/client.key
+    secret "istio-ingressgateway-certs" created
+    secret "istio-ingressgateway-ca-certs" created
+    secret "alb-certs" created
```
maybe a follow up blog - would be good to see steps with SDS gateway.

vadimeisenbergibm replied:
I can write a follow-up blog, but it is a pretty simple transformation from file-mounted certificates to the SDS provided one.

linsun replied:
I think it would be good, since there are a lot of interest with SDS gateway.

linsun commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,389 @@
+1.  Delete the Istio Ingress Gateway's pod to reload the certificates:
+
+    {{< text bash >}}
+    $ kubectl delete pod -l istio=ingressgateway -n istio-system
```
this is where SDS makes it easier.

vadimeisenbergibm replied:
right

linsun commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,389 @@
+    (the `--cert` option) and the private key (the `--key` option):
+
+    {{< text bash >}}
+    $ curl https://$INGRESS_GATEWAY_DOMAIN/status/418 --cert alb_certs/client.crt  --key alb_certs/client.key
```
is user supposed to run the curl cmd from their laptop or ssh into a pod?

vadimeisenbergibm replied:
from the laptop, to access the service through the ingress gateway

linsun replied:
SG

linsun commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,389 @@
+        `"""`
+    {{< /text >}}
+
+Congratulations! You configured the IKS Ingress ALB to securely send traffic to Istio ingress gateway.
```
I think it would be good to recap and explain how the traffic is mTLS among the IKS ALB to istio ingress gateway.  Basically, for istio ingress gateway, the istio ingress gateway is configured with certs/keys from the ALB domain.  For ALB, user has to append the issuer certificate of [Let's Encrypt](https://letsencrypt.org) to the certificate of ingress gateway so the ingress gateway can validate key/certs from ALB.  Further, from the ALB you direct traffic to istio-ingress gateway at path /status.

vadimeisenbergibm replied:
Done in https://github.com/istio/istio.io/pull/5098/commits/e6307814cc8f1a640f6ad8461994f9573c018f46

linsun commented:
@vadimeisenbergibm I took another pass today, I think it is mostly ready to go... really long blog but you articulated the steps well.

Just a few comments, then it would be ready from my perspective.

Would appreciate @frankbu to review it too.

vadimeisenbergibm commented:
Thank you @linsun, for the review, I addressed your comments. Would love to get feedback from @frankbu as well.

frankbu commented:
@vadimeisenbergibm I think you need to rebase your branch to get the netlify preview to build.

vadimeisenbergibm commented:
@frankbu I have fixed the netlify.

frankbu commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,394 @@
+keywords: [traffic-management,ingress,file-mount-credentials,iks]
+---
+
+In this blog post I show how to configure the [Ingress Application Load Balancer (ALB)](https://cloud.ibm.com/docs/containers?topic=containers-ingress-about)
```
This overview section is packed with information, but I don't think it's doing a good job of explaining why users would want to set up this relatively complicated configuration. What is the goal here? If I'm understanding correctly, it is showing how to securely configure an application with a single entry point (the ALB domain name) where the services implementing the application are a mix of Istio and non-Istio services. Otherwise, if all the services are all Istio enabled, users would just use the gateway domain name, with no need for the ALB at all. Is that correct? I think it sort of eludes to that, but it doesn't say that clearly. If I'm understanding this correctly, I think even the title should be more along those lines (e.g., "Using an IKS ALB to securely access a partially Istio-enabled application".

vadimeisenbergibm replied:
@frankbu 

> Otherwise, if all the services are all Istio enabled, users would just use the gateway domain name, with no need for the ALB at all.

Or the users still need some features of ALB and want to continue using it, even for Istio-enabled services. The blog post is intended for the users of IKS Ingress (ALB), they would know why they need to use ALB, which features they want.

rvennam replied:
Right - users want to continue using the default IKS nginx based ALB because it gives them features via annotations that are not available with Istio Gateways. Chaining the ALB -> Istio Gateway -> Istio enabled apps allows them in theory to have the best of both worlds, but they need to know how to enable mTLS between the ALB pods and the Istio gateway pods

adammil2000 commented:
@GregHanson You are a requested reviewer for this PR. Can you confirm your review/approval for it? Thanks!

rvennam commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,394 @@
+ALB. This ingress-traffic routing is configured using a Kubernetes
+[Ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/) resource with
+[ALB-specific annotations](https://cloud.ibm.com/docs/containers?topic=containers-ingress_annotation). IKS provides a
+DNS domain name for each cluster. The domain name is in the following format: `<cluster_name>.<region>.containers.appdomain.cloud`, for example
```
```suggestion
DNS domain name for each cluster 
```

rvennam commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,394 @@
+[Ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/) resource with
+[ALB-specific annotations](https://cloud.ibm.com/docs/containers?topic=containers-ingress_annotation). IKS provides a
+DNS domain name for each cluster. The domain name is in the following format: `<cluster_name>.<region>.containers.appdomain.cloud`, for example
+`mycluster.us-south.containers.appdomain.cloud`. In addition to registering a DNS domain, IKS provides a TLS certificate
```
```suggestion
 a TLS certificate
```

rvennam commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,394 @@
+that matches the domain, and a private key for the certificate. IKS stores the certificates and the private key in a
+[Kubernetes secret](https://kubernetes.io/docs/concepts/configuration/secret/).
+
+When you start using Istio in your IKS cluster, the recommended way is to use the
```
```suggestion
When you start using Istio in your IKS cluster, the recommended method to send traffic to your Istio enabled workloads is by using the
```

rvennam commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,394 @@
+[Istio Ingress Gateway](/docs/tasks/traffic-management/ingress/ingress-control/) instead of using the
+[Kubernetes Ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/). One of the main reasons to use
+Istio ingress gateway is the fact the ALB provided by IKS will not be able to
+communicate directly with the services inside the mesh when you enable Istio mutual TLS. During your transition to
```
```suggestion
communicate directly with the services inside the mesh when you enable STRICT mutual TLS. During your transition to
```

rvennam commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,394 @@
+having only Istio ingress gateway as your main entry point, you can continue to use the traditional Ingress for
+non-Istio services while using the Istio ingress gateway for services that are part of the mesh.
+
+IKS provides a convenient way for clients to access Istio ingress gateway by letting you
```
This paragraph should be removed IMO as it talks about skipping the ALB and can distract the reader.

rvennam replied:
OK, I see why you talk about this. The instructions leverage both DNS entries. Maybe add a couple of sentences describing what the instructions are doing with these two DNS and their secrets?

rvennam commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,394 @@
+IP with an IKS command. The domain is in the following [format](https://cloud.ibm.com/docs/containers?topic=containers-loadbalancer_hostname#loadbalancer_hostname_format): `<cluster_name>-<globally_unique_account_HASH>-0001.<region>.containers.appdomain.cloud`, for example `mycluster-a1b2cdef345678g9hi012j3kl4567890-0001.us-south.containers.appdomain.cloud`. In the same way as for the ALB domain, ...
+storing them in another Kubernetes secret.
+
+In this blog post you configure the IKS Ingress ALB to direct traffic to the services inside an Istio service mesh
```
```suggestion
This blog describes how you can configure the IKS Ingress ALB to direct traffic to the services inside an Istio service mesh
```

rvennam commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,394 @@
+    deployment.apps/httpbin created
+    {{< /text >}}
+
+1.  Enforce mutual TLS for the traffic to the services in
```
Suggest to remove creating this Policy and DR for httptools
1. Keep blog focused on the ALB -> Gateway
2. Policy not used in 1.5
3. DR not required with automtls

vadimeisenbergibm replied:
Fixed in https://github.com/istio/istio.io/pull/5098/commits/a99af4d702bece9f8749b89233366d345b5478b0. Enforcing mutual TLS was done to show why you need the traffic to go through the Istio Ingress Gateway - because of the mutual TLS. But, I guess, we can just explain it without actually enforcing it.

rvennam commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,394 @@
+1.  Create a DNS domain for the IP of the Istio Ingress Gateway service:
+
+    {{< text bash >}}
+    $ ibmcloud ks nlb-dns-create --cluster $CLUSTER_NAME --ip $INGRESS_GATEWAY_IP
```
```suggestion
    $ ibmcloud ks nlb-dns create classic --cluster $CLUSTER_NAME --ip $INGRESS_GATEWAY_IP
```

rvennam commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,394 @@
+1.  Define a `Gateway` to allow access on port 443 only, with mutual TLS:
+
+    {{< text bash >}}
+    $ kubectl apply -f - <<EOF
```
```suggestion
    $ kubectl apply -n httptools -f - <<EOF
```

rvennam commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,394 @@
+1. Configure routes for traffic entering via the `Gateway`:
+
+    {{< text bash >}}
+    $ kubectl apply -f - <<EOF
```
```suggestion
    $ kubectl apply -n httptools -f - <<EOF
```

rvennam commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,394 @@
+    {{< /text >}}
+
+Congratulations! You configured IKS Ingress ALB to securely send traffic to Istio ingress gateway. You allocated a host
+name and certificate for your Istio ingress gateway and used that certificate as the server certificate for Istio
```
This is key information that should be described in the introduction as well.

rvennam commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,394 @@
+## Configure ALB
+
+You need to configure your Ingress resource to direct traffic to the Istio Ingress gateway while using the
+certificate stored in the `alb-certs` secret.
```
```suggestion
certificate stored in the `alb-certs` secret. Normally, the ALB decrypts HTTPS requests before forwarding traffic to your apps. We can configure the ALB to re-encrypt the traffic before it is forwarded to the Istio ingress gateway by using the `ssl-services` annotation on the Ingress resource. This annotation also allows us to specify the certificate stored in the `alb-certs` secret, required for mutual TLS.
```

rvennam commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,394 @@
+storing them in another Kubernetes secret.
+
+In this blog post you configure the IKS Ingress ALB to direct traffic to the services inside an Istio service mesh
+through the Istio ingress gateway, while using mutual TLS authentication between ALB and the gateway. For the mutual TLS authentication, you configure ALB and the Istio ingress gateway to use the certificates and the keys provided by IKS.
```
```suggestion
through the Istio ingress gateway, while using mutual TLS authentication between ALB and the gateway. For the mutual TLS authentication, you will configure ALB and the Istio ingress gateway to use the certificates and keys provided by IKS for ALB and NLB host names. You will use the NLB certificate as the server certificate for Istio ingress gateway and use the ALB certificate ALB as the client certificate of ALB . You then use a
[Let's Encrypt](https://letsencrypt.org) certificate as the root for both client and server certificates.
```

Rachael-Graham replied:
make sure to add `the` before ALB (multiple instances of this throughout this blog), so always `the ALB` or `an ALB` etc

vadimeisenbergibm replied:
Fixed by https://github.com/istio/istio.io/pull/5098/commits/7adfa2fbb2fa2ce5f221c3ea05cfe12bc612c1a6.

rvennam commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,394 @@
+non-Istio services while using the Istio ingress gateway for services that are part of the mesh.
+
+IKS provides a convenient way for clients to access Istio ingress gateway by letting you
+[register a DNS domain](https://cloud.ibm.com/docs/containers?topic=containers-loadbalancer_hostname) for the gateway's
```
```suggestion
[register a new DNS subdomain](https://cloud.ibm.com/docs/containers?topic=containers-loadbalancer_hostname) for the Istio ingress gateway's NLB 
```

vadimeisenbergibm replied:
IKS allows registering a new DNS subdomain for any load balancer/gateway (IKS calls such a load balancer/gateway "NLB"). So Istio ingress gateway is a kind of such "NLB", so NLB is redundant here and it is rather IKS-specific term anyway.

vadimeisenbergibm replied:
Addressed by https://github.com/istio/istio.io/pull/5098/commits/5c88b59a3df44e4402a5c9fc546e056e8e5f8da7 and https://github.com/istio/istio.io/pull/5098/commits/858333f67cb58829beef92a8c923259903f4bdec . Left out NLB, the point of the sentence is that IKS allows you to register a new subdomain for the IP address of the Istio Ingress Gateway. It is just the general using of subdomains: you register a subdomain for an IP address, so you will be able to use that subdomain instead of typing the IP address.

jpapejr commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,394 @@
+    ...
+    {{< /text >}}
+
+    Wait until the status of the certificate (the third field) in the line that matches the IP of the Istio ingress
```
Believe this is actually the fourth field, not the third: 
```
        1.          2.               3.                        4.                            5.                                6. 
Hostname,IP(s),Health Monitor,SSL Cert Status,SSL Cert Secret Name,Secret Namespace 
```

vadimeisenbergibm replied:
Fixed by https://github.com/istio/istio.io/pull/5098/commits/29a21c0cef9de5b6c20d1f4eeb555c1b40be7d81.

Rachael-Graham commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,394 @@
+title: Securely direct traffic from IBM Cloud Kubernetes Service Ingress to Istio Ingress Gateway
+description: Configure the IBM Cloud Kubernetes Service Application Load Balancer to direct traffic to Istio Ingress gateway with mutual TLS.
+subtitle: Configure the IBM Cloud Kubernetes Service Application Load Balancer to direct traffic to Istio Ingress gateway with mutual TLS
+publishdate: 2019-11-07
```
does this need to be changed?

vadimeisenbergibm replied:
Fixed by https://github.com/istio/istio.io/pull/5098/commits/29a21c0cef9de5b6c20d1f4eeb555c1b40be7d81.

vadimeisenbergibm replied:
And also by https://github.com/istio/istio.io/pull/5098/commits/e93de4dad9242e433c149eccd1a267e8f252604d.

Rachael-Graham commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,394 @@
+When you start using Istio in your IKS cluster, the recommended way is to use the
+[Istio Ingress Gateway](/docs/tasks/traffic-management/ingress/ingress-control/) instead of using the
+[Kubernetes Ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/). One of the main reasons to use
+Istio ingress gateway is the fact the ALB provided by IKS will not be able to
```
```suggestion
the Istio ingress gateway is the fact the ALB provided by IKS will not be able to
```

Rachael-Graham commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,394 @@
+Using certificates provided by IKS saves you the overhead of managing your own certificates for the connection between ALB and
+the Istio ingress gateway.
+
+The traffic to the services without Istio sidecar can continue to flow as before, directly from ALB.
```
```suggestion
Traffic to the services without an Istio sidecar can continue to flow as before directly from the ALB.
```

Rachael-Graham commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,394 @@
+The traffic to the services without Istio sidecar can continue to flow as before, directly from ALB.
+
+The diagram below exemplifies the described setting. It shows two services in the cluster, `service A` and `service B`.
+`service A` has Istio sidecar injected and it requires mutual TLS. `service B` has no Istio sidecar. `service B` can
```
```suggestion
`service A` has an Istio sidecar injected and requires mutual TLS. `service B` has no Istio sidecar. `service B` can
```

Rachael-Graham commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,394 @@
+
+The diagram below exemplifies the described setting. It shows two services in the cluster, `service A` and `service B`.
+`service A` has Istio sidecar injected and it requires mutual TLS. `service B` has no Istio sidecar. `service B` can
+be accessed by clients through the ALB, while ALB directly communicates with `service B`. `service A` can be also
```
```suggestion
be accessed by clients through the ALB, which directly communicates with `service B`. `service A` can be also
```

Rachael-Graham commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,394 @@
+## Create secrets for ALB and Istio ingress gateway
+
+IKS generates a TLS certificate and a private key and stores them as a secret in the default namespace when you register
+a DNS domain for an external IP by the `ibmcloud ks nlb-dns-create` command. IKS stores the Application Load Balancer's
```
```suggestion
a DNS domain for an external IP by using the `ibmcloud ks nlb-dns-create` command. IKS stores the ALB's
```

Rachael-Graham commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,394 @@
+IKS generates a TLS certificate and a private key and stores them as a secret in the default namespace when you register
+a DNS domain for an external IP by the `ibmcloud ks nlb-dns-create` command. IKS stores the Application Load Balancer's
+certificate and private key also as a secret in the default namespace. You need these credentials to establish the
+identities the ALB and the Istio ingress gateway will present one to another during mutual TLS authentication between
```
```suggestion
identities that the ALB and the Istio ingress gateway will present during the mutual TLS authentication between
```

Rachael-Graham commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,394 @@
+certificate and private key also as a secret in the default namespace. You need these credentials to establish the
+identities the ALB and the Istio ingress gateway will present one to another during mutual TLS authentication between
+them. You configure the ALB and the Istio ingress gateway to exchange these certificates signed by their private keys,
+and to trust the certificates one of another.
```
```suggestion
and to trust the certificates of one another.
```

rvennam commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,394 @@
+    {{< text bash >}}
+    $ kubectl delete ingress alb-ingress -n istio-system
+    $ kubectl delete virtualservice default-ingress
+    $ kubectl delete gateway default-ingress-gateway
```
```suggestion
    $ kubectl delete gateway default-ingress-gateway -n httptools
```

rvennam commented on content/en/blog/2019/alb-ingress-gateway-iks/index.md:
```diff
@@ -0,0 +1,394 @@
+
+    {{< text bash >}}
+    $ kubectl delete ingress alb-ingress -n istio-system
+    $ kubectl delete virtualservice default-ingress
```
```suggestion
    $ kubectl delete virtualservice default-ingress -n httptools
```

frankbu commented:
/retest

istio-testing commented:
@vadimeisenbergibm: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
k8s-tests_istio.io | a24e453e0ad910ea746ee880d872fbfaca66d4f1 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio.io/5098/k8s-tests_istio.io/1793) | `/test k8s-tests_istio.io`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->
